src/github.com/matrix-org/dendrite/mediaapi/writers/fileutils.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2017 Vector Creations Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writers
```

Comment:
I don't like how this file is inside the `writers` package. It's too big now. It means there's way too many functions I need to keep track of which are in scope for the uploads/downloads file.

At call-sites in the upload code for example, we use both `storeFileAndMetadata` and `readAndHashAndWriteWithLimit` and I keep failing to know where I need to look for the impl. I know the conceptual difference and can work it out if I think about it for a bit, but this is needless cognitive effort on my part just to know which file to look in.

Please can you move this to a `fileutils` package. That will keep things much clearer at call-sites then (since you'll need to do `fileutils.readAndHashAndWriteWithLimit` vs just `storeFileAndMetadata`). It'll also keep you honest with regards to only putting file-related functionality in this file.